internal/crypto/certgen.go

```diff
@@ -0,0 +1,278 @@
+		if err != nil {
+			return nil, err
+		}
+
+		var egDNSNames, envoyDNSNames []string
+		egProvider := egCfg.GetProvider().Type
+		switch egProvider {
+		case v1alpha1.ProviderTypeKubernetes:
+			egDNSNames = kubeServiceNames(DefaultEnvoyGatewayDNSPrefix, DefaultNamespace, DefaultDNSSuffix)
+			envoyDNSNames = kubeServiceNames(DefaultEnvoyDNSPrefix, DefaultNamespace, DefaultDNSSuffix)
+		default:
+			// Kubernetes is the only supported Envoy Gateway provider.
+			return nil, fmt.Errorf("unsupported provider type %v", egProvider)
+		}
+
+		egCertReq := &certificateRequest{
+			caCertPEM:  caCertPEM,
+			caKeyPEM:   caKeyPEM,
+			expiry:     expiry,
+			commonName: DefaultEnvoyGatewayDNSPrefix,
+			altNames:   egDNSNames,
+		}
+
+		egCert, egKey, err := newCert(egCertReq)
+		if err != nil {
+			return nil, err
+		}
+
+		envoyCertReq := &certificateRequest{
+			caCertPEM:  caCertPEM,
+			caKeyPEM:   caKeyPEM,
+			expiry:     expiry,
+			commonName: DefaultEnvoyDNSPrefix,
+			altNames:   envoyDNSNames,
+		}
+
+		envoyCert, envoyKey, err := newCert(envoyCertReq)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Certificates{
+			CACertificate:           caCertPEM,
+			EnvoyGatewayCertificate: egCert,
+			EnvoyGatewayPrivateKey:  egKey,
+			EnvoyCertificate:        envoyCert,
+			EnvoyPrivateKey:         envoyKey,
+		}, nil
+	default:
+		// Envoy Gateway, e.g. self-signed CA, is the only supported certificate provider.
+		return nil, fmt.Errorf("unsupported certificate provider type %v", certCfg.Provider.Type)
+	}
+}
+
+// newCert generates a new keypair based on the given the request.
+// The return values are cert, key, err.
+func newCert(request *certificateRequest) ([]byte, []byte, error) {
+
+	caKeyPair, err := tls.X509KeyPair(request.caCertPEM, request.caKeyPEM)
+	if err != nil {
+		return nil, nil, err
+	}
+	caCert, err := x509.ParseCertificate(caKeyPair.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	caKey, ok := caKeyPair.PrivateKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("CA private key has unexpected type %T", caKeyPair.PrivateKey)
+	}
+
+	newKey, err := rsa.GenerateKey(rand.Reader, keySize)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot generate key: %v", err)
+	}
+
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: newSerial(now),
+		Subject: pkix.Name{
+			CommonName: request.commonName,
+		},
+		NotBefore:    now.UTC().AddDate(0, 0, -1),
+		NotAfter:     request.expiry.UTC(),
+		SubjectKeyId: bigIntHash(newKey.N),
+		KeyUsage: x509.KeyUsageDigitalSignature |
+			x509.KeyUsageDataEncipherment |
+			x509.KeyUsageKeyEncipherment |
+			x509.KeyUsageContentCommitment,
+		DNSNames: request.altNames,
+	}
+	newCert, err := x509.CreateCertificate(rand.Reader, template, caCert, &newKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newKeyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(newKey),
```

Comment:
Getting in to bike-shed territory, but any reason to not use `"PRIVATE KEY"` and `x509.MarshalPKCS8PrivateKey`? It'd have fewer things to change if we ever want to switch away from RSA.